In [2]:
import pandas as pd
import numpy as np
pd.set_option('display.max_rows', 1000) 
pd.set_option('display.max_columns', 1000) 

In [3]:
def basic_details(df):
    """Returns basic statistics for dataset - duplicates,
    % of zero, nan, unique values, min, max, mean, median and standard deviation."""
    b = pd.DataFrame()
    b['Missing, %'] = round(df.isnull().sum()/df.shape[0]*100, 2)
    b['Zero, %'] = round(df.isin([0]).sum() /df.shape[0]*100, 2)
    b['N unique value'] = df.nunique()
    b['dtype'] = df.dtypes
    numeric_cols = df.select_dtypes(include=[np.number]).columns
    b.loc[numeric_cols, 'Min'] = df[numeric_cols].min().round(2)
    b.loc[numeric_cols, '1%'] = df[numeric_cols].quantile(0.01)
    b.loc[numeric_cols, '25%'] = df[numeric_cols].quantile(0.25)
    b.loc[numeric_cols, 'Mean'] = df[numeric_cols].mean().round(2)
    b.loc[numeric_cols, 'Median'] = df[numeric_cols].median().round(2)
    b.loc[numeric_cols, '75%'] = df[numeric_cols].quantile(0.75)
    b.loc[numeric_cols, '995%'] = df[numeric_cols].quantile(0.995)
    b.loc[numeric_cols, 'Max'] = df[numeric_cols].max().round(2)
    b.loc[numeric_cols, 'Std'] = df[numeric_cols].std().round(2)
    return b

In [4]:
# load train and test datasets
df_train = pd.read_csv(r'F:\DS\Kernel\data\train.csv')
df_test = pd.read_csv(r'F:\DS\Kernel\data\test.csv')
df_train.shape, df_test.shape

((5734, 162), (2192, 162))

In [5]:
# checking for dplicates
df_train.duplicated().sum(), df_test.duplicated().sum()

(47, 0)

In [6]:
# need full duplicates to be dropped
df_train = df_train.drop_duplicates()

In [7]:
# take a close look at data and check for duplicates by Field and Year features
df_train[df_train.duplicated(subset=['Field', 'Year'], keep=False)].sort_values(by='Field').head(6)

Field  Year  Cluster     Area       Yield  Geozone  Predecessor    FAO   
1325    203  2016        3   58.610  100.684400       32            3  280.0  \
1324    203  2016        3   58.610  100.684400       32            1  280.0   
5088    217  2018        6   61.471  129.365327       19            4  270.0   
5087    217  2018        6   61.471  129.365327       19            2  270.0   
5090    219  2018        6  162.252  144.030545       19            2    NaN   
5089    219  2018        6  162.252  144.030545       19            1    NaN   

         N_kg/ga    P_kg/ga    K_kg/ga  5_YEARLY_median_AVERAGE_TEMPERATURE   
1325  173.735813  36.831599  17.326395                            16.157020  \
1324  173.735813  36.831599  17.326395                            16.157020   
5088  170.616391  54.224594  68.028339                            15.912748   
5087  170.616391  54.224594  68.028339                            15.912748   
5090  134.320967  54.904420  69.771220                            15.912748   
5089  134.320967  54.904420  69.771220                            15.912748   

      6_YEARLY_median_AVERAGE_TEMPERATURE   
1325                            19.990389  \
1324                            19.990389   
5088                            19.874560   
5087                            19.874560   
5090                            19.874560   
5089                            19.874560   

      7_YEARLY_median_AVERAGE_TEMPERATURE   
1325                            21.764953  \
1324                            21.764953   
5088                            21.582069   
5087                            21.582069   
5090                            21.582069   
5089                            21.582069   

      8_YEARLY_median_AVERAGE_TEMPERATURE   
1325                            21.200779  \
1324                            21.200779   
5088                            20.920000   
5087                            20.920000   
5090                            20.920000   
5089                            20.920000   

      9_YEARLY_median_AVERAGE_TEMPERATURE  5_YEARLY_median_EVAPORATION   
1325                            15.738342                         4.03  \
1324                            15.738342                         4.03   
5088                            15.504203                         3.68   
5087                            15.504203                         3.68   
5090                            15.504203                         3.68   
5089                            15.504203                         3.68   

      6_YEARLY_median_EVAPORATION  7_YEARLY_median_EVAPORATION   
1325                        4.885                         4.75  \
1324                        4.885                         4.75   
5088                        4.270                         4.30   
5087                        4.270                         4.30   
5090                        4.270                         4.30   
5089                        4.270                         4.30   

      8_YEARLY_median_EVAPORATION  9_YEARLY_median_EVAPORATION   
1325                         4.01                        2.305  \
1324                         4.01                        2.305   
5088                         3.72                        2.260   
5087                         3.72                        2.260   
5090                         3.72                        2.260   
5089                         3.72                        2.260   

      5_YEARLY_median_HUMIDITY(%)  6_YEARLY_median_HUMIDITY(%)   
1325                         60.0                         60.0  \
1324                         60.0                         60.0   
5088                         61.0                         63.0   
5087                         61.0                         63.0   
5090                         61.0                         63.0   
5089                         61.0                         63.0   

      7_YEARLY_median_HUMIDITY(%)  

In [8]:
# lets take Filed with max value_counts() value
df_train[df_train.Field==1147]

Field  Year  Cluster     Area       Yield  Geozone  Predecessor    FAO   
2172   1147  2018        4  110.680  162.553126        6            1  350.0  \
2173   1147  2018        4  110.680  162.553126        6            4  350.0   
2174   1147  2018        4  110.680  162.553126        6            2  350.0   
2175   1147  2019        4  103.338   93.731870        6            0  330.0   
2176   1147  2016        4   82.490  134.411600        6            0    NaN   
2177   1147  2016        4   82.490  134.411600        6            4    NaN   
2179   1147  2016        4   82.490  134.411600        6            1    NaN   

         N_kg/ga    P_kg/ga    K_kg/ga  5_YEARLY_median_AVERAGE_TEMPERATURE   
2172  129.173338  53.052123  72.680882                             15.90439  \
2173  129.173338  53.052123  72.680882                             15.90439   
2174  129.173338  53.052123  72.680882                             15.90439   
2175  183.576226  52.030105  49.539956                             15.90439   
2176  177.765499  34.760280  -8.559050                             15.90439   
2177  177.765499  34.760280  -8.559050                             15.90439   
2179  177.765499  34.760280  -8.559050                             15.90439   

      6_YEARLY_median_AVERAGE_TEMPERATURE   
2172                            19.762328  \
2173                            19.762328   
2174                            19.762328   
2175                            19.762328   
2176                            19.762328   
2177                            19.762328   
2179                            19.762328   

      7_YEARLY_median_AVERAGE_TEMPERATURE   
2172                            21.516067  \
2173                            21.516067   
2174                            21.516067   
2175                            21.516067   
2176                            21.516067   
2177                            21.516067   
2179                            21.516067   

      8_YEARLY_median_AVERAGE_TEMPERATURE   
2172                            20.766404  \
2173                            20.766404   
2174                            20.766404   
2175                            20.766404   
2176                            20.766404   
2177                            20.766404   
2179                            20.766404   

      9_YEARLY_median_AVERAGE_TEMPERATURE  5_YEARLY_median_EVAPORATION   
2172                            15.407706                         3.61  \
2173                            15.407706                         3.61   
2174                            15.407706                         3.61   
2175                            15.407706                         3.61   
2176                            15.407706                         3.61   
2177                            15.407706                         3.61   
2179                            15.407706                         3.61   

      6_YEARLY_median_EVAPORATION  7_YEARLY_median_EVAPORATION   
2172                        4.155                         4.11  \
2173                        4.155                         4.11   
2174                        4.155                         4.11   
2175                        4.155                         4.11   
2176                        4.155                         4.11   
2177                        4.155                         4.11   
2179                        4.155                         4.11   

      8_YEARLY_median_EVAPORATION  9_YEARLY_median_EVAPORATION   
2172                          3.5                        2.085  \
2173                          3.5                        2.085   
2174                          3.5                        2.085   
2175                          3.5                        2.085   
2176                          3.5                        2.085   
2177                          3.5                        2.085   
2179                          3.5                        2.085   

In [9]:
# checking the same at test dataset
df_test[df_test.duplicated(subset=['Field', 'Year'], keep=False)].sort_values(by='Field').head(6)

Field  Year  Cluster     Area  Yield  Geozone  Predecessor    FAO   
133   2667  2020        1  104.653      0       14            1  330.0  \
134   2667  2020        1  104.653      0       14            4  330.0   
169   2708  2020        1  153.379      0       14            1  320.0   
170   2708  2020        1  153.379      0       14            4  320.0   
916   2835  2020        5  118.515      0       16            0  320.0   
917   2835  2020        5  118.515      0       16            4  320.0   

        N_kg/ga    P_kg/ga    K_kg/ga  5_YEARLY_median_AVERAGE_TEMPERATURE   
133  229.341825  61.405922  85.991897                            15.523175  \
134  229.341825  61.405922  85.991897                            15.523175   
169  206.316535  56.729970  73.195744                            15.523175   
170  206.316535  56.729970  73.195744                            15.523175   
916  124.879281  41.223734  -7.117356                            16.365038   
917  124.879281  41.223734  -7.117356                            16.365038   

     6_YEARLY_median_AVERAGE_TEMPERATURE  7_YEARLY_median_AVERAGE_TEMPERATURE   
133                            19.241317                            21.013416  \
134                            19.241317                            21.013416   
169                            19.241317                            21.013416   
170                            19.241317                            21.013416   
916                            20.237856                            21.860072   
917                            20.237856                            21.860072   

     8_YEARLY_median_AVERAGE_TEMPERATURE  9_YEARLY_median_AVERAGE_TEMPERATURE   
133                            20.258808                            15.103848  \
134                            20.258808                            15.103848   
169                            20.258808                            15.103848   
170                            20.258808                            15.103848   
916                            21.480000                            15.930690   
917                            21.480000                            15.930690   

     5_YEARLY_median_EVAPORATION  6_YEARLY_median_EVAPORATION   
133                         3.36                        3.905  \
134                         3.36                        3.905   
169                         3.36                        3.905   
170                         3.36                        3.905   
916                         4.11                        4.965   
917                         4.11                        4.965   

     7_YEARLY_median_EVAPORATION  8_YEARLY_median_EVAPORATION   
133                         3.94                         3.25  \
134                         3.94                         3.25   
169                         3.94                         3.25   
170                         3.94                         3.25   
916                         4.91                         4.10   
917                         4.91                         4.10   

     9_YEARLY_median_EVAPORATION  5_YEARLY_median_HUMIDITY(%)   
133                        2.030                         65.0  \
134                        2.030                         65.0   
169                        2.030                         65.0   
170                        2.030                         65.0   
916                        2.285                         58.0   
917                        2.285                         58.0   

     6_YEARLY_median_HUMIDITY(%)  7_YEARLY_median_HUMIDITY(%)   
133                         68.0                         67.0  \
134                         68.0                         67.0   
169                         68.0                         67.0   
170                         68.0                         67.0   
916                         57.0                         59.0   
917                         57.0             

## Observations

### Multiple Predecessors
This observation indicates that a given culture in the Field may have several Predecessors, and each of them is described in the dataset with duplicate rows, specifying the predecessors. Additional processing will be required during modeling to handle this case.

### Incomplete Data Across Years
Another observation is that for each Field, there may not necessarily be observations for all 4 years that are presented in the dataset.


In [10]:
basic_details(df_train)

Missing, %  Zero, %  N unique value   
Field                                      0.00     0.02            3684  \
Year                                       0.00     0.00               4   
Cluster                                    0.00    41.01               7   
Area                                       0.00     0.00            5149   
Yield                                      0.00     0.00            5401   
Geozone                                    0.00     2.25              42   
Predecessor                                0.00    35.43               5   
FAO                                       13.73     0.00              22   
N_kg/ga                                    0.00     0.00            5561   
P_kg/ga                                    3.92     0.00            5343   
K_kg/ga                                    5.79     0.00            5242   
5_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              64   
6_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              64   
7_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              64   
8_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              63   
9_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              64   
5_YEARLY_median_EVAPORATION                0.00     0.00              43   
6_YEARLY_median_EVAPORATION                0.00     0.00              55   
7_YEARLY_median_EVAPORATION                0.00     0.00              50   
8_YEARLY_median_EVAPORATION                0.00     0.00              48   
9_YEARLY_median_EVAPORATION                0.00     0.00              48   
5_YEARLY_median_HUMIDITY(%)                0.00     0.00              17   
6_YEARLY_median_HUMIDITY(%)                0.00     0.00              16   
7_YEARLY_median_HUMIDITY(%)                0.00     0.00              15   
8_YEARLY_median_HUMIDITY(%)                0.00     0.00              23   
9_YEARLY_median_HUMIDITY(%)                0.00     0.00              17   
5_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              64   
6_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              64   
7_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              64   
8_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              64   
9_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              64   
5_YEARLY_sum_RAINFALL                      0.00     0.00              64   
6_YEARLY_sum_RAINFALL                      0.00     0.00              54   
7_YEARLY_sum_RAINFALL                      0.00     0.00              55   
8_YEARLY_sum_RAINFALL                      0.00     0.00              55   
9_YEARLY_sum_RAINFALL                      0.00     0.00              52   
22_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             181   
23_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             172   
24_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             165   
25_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             176   
26_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             163   
27_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             177   
28_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             174   
29_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             178   
30_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             174   
31_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             179   
32_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             178   
33_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             181   
34_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             177   
35_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             174   
36_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             177   
37_WEEK_AVERAGE_TEMPERATURE                0.00     0.00             171   
38_WEEK_A

In [11]:
basic_details(df_test)

Missing, %  Zero, %  N unique value   
Field                                      0.00     0.05            2187  \
Year                                       0.00     0.00               1   
Cluster                                    0.00    49.27               7   
Area                                       0.00     0.00            2176   
Yield                                      0.00   100.00               1   
Geozone                                    0.00     0.64              41   
Predecessor                                0.00    40.37               5   
FAO                                       22.86     0.00              11   
N_kg/ga                                    0.00     0.00            2187   
P_kg/ga                                   10.45     0.00            1958   
K_kg/ga                                   10.54     0.00            1956   
5_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              57   
6_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              57   
7_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              57   
8_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              57   
9_YEARLY_median_AVERAGE_TEMPERATURE        0.00     0.00              57   
5_YEARLY_median_EVAPORATION                0.00     0.00              36   
6_YEARLY_median_EVAPORATION                0.00     0.00              51   
7_YEARLY_median_EVAPORATION                0.00     0.00              44   
8_YEARLY_median_EVAPORATION                0.00     0.00              46   
9_YEARLY_median_EVAPORATION                0.00     0.00              46   
5_YEARLY_median_HUMIDITY(%)                0.00     0.00              15   
6_YEARLY_median_HUMIDITY(%)                0.00     0.00              17   
7_YEARLY_median_HUMIDITY(%)                0.00     0.00              15   
8_YEARLY_median_HUMIDITY(%)                0.00     0.00              23   
9_YEARLY_median_HUMIDITY(%)                0.00     0.00              14   
5_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              57   
6_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              57   
7_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              57   
8_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              57   
9_YEARLY_sum_CALCULATED_RADIATION          0.00     0.00              57   
5_YEARLY_sum_RAINFALL                      0.00     0.00              57   
6_YEARLY_sum_RAINFALL                      0.00     0.00              48   
7_YEARLY_sum_RAINFALL                      0.00     0.00              50   
8_YEARLY_sum_RAINFALL                      0.00     0.00              49   
9_YEARLY_sum_RAINFALL                      0.00     0.00              49   
22_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              50   
23_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              44   
24_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              45   
25_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              53   
26_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              40   
27_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              35   
28_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              49   
29_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              50   
30_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              47   
31_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              48   
32_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              49   
33_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              38   
34_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              53   
35_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              49   
36_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              51   
37_WEEK_AVERAGE_TEMPERATURE                0.00     0.00              41   
38_WEEK_A

# Descriptive Statistics Analysis Report

## Conclusions obtained from the analysis of descriptive statistics for each feature for the train set:

### General Characteristics:
- **Field - Field ID**: No missing values, from 0 to 4180 (3684 unique values).
- **Year - Observation Year**: No missing values, from 2016 to 2019.
- **Cluster - Cluster ID**: No missing values, from 0 to 6 (7 unique values).

### Area and Yield:
- **Area - Field area in hectares**: No missing values, from 30 to 730 (a significant increase in value over a very small interval between 99.5 and 100 percentile, which is often an indicator of the presence of outliers).
- **Yield - Yield in ton per hectare**: No missing values, from 3.4 to 190.4 (close to normally distributed without outliers).

### Geozone and Predecessor:
- **Geozone - Geozone ID**: No missing values, from 0 to 41.
- **Predecessor - Predecessor ID (type of crop that was grown in a given field in the previous year)**: No missing values, from 0 to 4.

### FAO Index:
- **FAO - Hybrid FAO**: 13.8% missing values, from 180 to 440 (normally distributed without outliers) - hypothesis of the presence of missing values - the FAO index for them is currently not defined or assigned. If necessary, need to replace the NaNs with the average FAO value for the Geozone (hypothesis: hybrids with similar FAO characteristics will be grown in geographical areas with similar climates).

### Fertilizers Applied (N, P, K):
- **'N_kg/ga'**: The amount of applied fertilizers in kg per hectare, Nitrogen, no missing values (nearly normal distribution without outliers).
- **'P_kg/ga'**: The amount of applied fertilizers in kg per hectare, Phosphorus, 4% of missing values (close to normal distribution, with outliers).
- **'K_kg/ga'**: The amount of applied fertilizers in kg per hectare, Potassium, 6% of missing values, scaled from -97 to 197. Need to rescale between 0 and 1 (check test set distribution).

### Weather and Vegetation Indices:
- **AVERAGE_TEMPERATURE**: Median temperatures on a 5-9 year basis, no missing values, normally distributed from 14.48 to 21.5 degrees.
- **EVAPORATION**: Median evaporation on a 5-9 year basis, no missing values, close to normal distribution.
- **HUMIDITY**: Median humidity on a 5-9 year basis, no missing values, normally distributed.
- **CALCULATED_RADIATION**: Sum calculated radiation on a 5-9 year basis, no missing values, from 5.9K to 15.4K units, close to normal distribution.
- **RAINFALL**: Sum rainfall on a 5-9 year basis, no missing values, from 299 to 1467 units, normally distributed.
- **WEEK_AVERAGE_TEMPERATURE, WEEK_CALCULATED_RADIATION, WEEK_EVAPORATION, WEEK_HUMIDITY**: No missing values, normally distributed.
- **WEEK_RAINFALL**: No missing values, not normal distribution, with outliers (the mean and median are far from each other, there is a large distance between the 99.5 and 100 percentiles).
- **WEEK_SUM_ACTIVE_TEMPERATURE**: No missing values, normally distributed.
- **Week_ndvi**: Value between 0 and 1, no missing values, close to normal distribution.

## Conclusions obtained from the analysis of descriptive statistics for the test set:

- **Field - Field ID**: No missing values, from 0 to 4180 (2187 unique values).
- **Year - Observation Year**: No missing values, 2020 year only.
- **Cluster - Cluster ID**: No missing values, from 0 to 6 (7 unique values).
- **Area - Field area in hectares**: No missing values, from 30 to 717 (a significant increase in value over a very small interval between 99.5 and 100 percentile, which is often an indicator of the presence of outliers).
- **Yield - Yield in ton per hectare**: Needs to be predicted.
- **Geozone - Geozone ID**: No missing values, from 0 to 42!! (instead of 41 in the train set).
- **Predecessor - Predecessor ID (type of crop that was grown in a given field in the previous year)**: No missing values, from 0 to 4.
- **FAO - Hybrid FAO (the higher the FAO value, the higher the yield of the hybrid)**: 22.9% missing values, from 250 to 390 (close to normal distribution without outliers) - hypothesis of the presence of missing values - the FAO index for them is currently not defined or assigned. If necessary, need to replace the NaNs with the average FAO value for the Geozone (hypothesis: hybrids with similar FAO characteristics will be grown in geographical areas with similar climates).
- **'N_kg/ga'**: The amount of applied fertilizers in kg per hectare, Nitrogen, no missing values (nearly normal distribution without outliers).
- **'P_kg/ga'**: The amount of applied fertilizers in kg per hectare, Phosphorus, 10.5% of missing values (close to normal distribution, with outliers).
- **'K_kg/ga'**: The amount of applied fertilizers in kg per hectare, Potassium, 10.5% of missing values, scaled feature -10 and 196. Need to rescale between 0 and 1 (check test set distribution).
- **AVERAGE_TEMPERATURE**: Median temperatures on a 5-9 year basis, no missing values, normally distributed.
- **EVAPORATION**: Median evaporation on a 5-9 year basis, no missing values, close to normal distribution.
- **HUMIDITY**: Median humidity on a 5-9 year basis, no missing values, normally distributed.
- **CALCULATED_RADIATION**: Sum calculated radiation on a 5-9 year basis, no missing values, close to normal distribution.
- **RAINFALL**: Sum rainfall on a 5-9 year basis, no missing values, normally distributed.
- **WEEK_AVERAGE_TEMPERATURE, WEEK_CALCULATED_RADIATION, WEEK_EVAPORATION, WEEK_HUMIDITY**: No missing values, normally distributed.
- **WEEK_RAINFALL**: No missing values, not normal distribution, with outliers (the mean and median are far from each other, there is a large distance between the 99.5 and 100 percentiles).
- **WEEK_SUM_ACTIVE_TEMPERATURE**: No missing values, normally distributed.
- **week_ndvi**: Value between 0 and 1, no missing values, close to normal distribution (**39_week_ndvi - all zero values**).

## Differences between test and train datasets: 
- **Year - Observation Year**: 2016-2019 vs 2020
- **Geozone - Geozone ID**: one additional geozone in test set
- **week_ndvi**: 39_week_ndvi - all zero values


## Notes and Considerations:

### Outlier Investigation:
- **Investigate outliers for the 'Area' feature in both datasets graphically**: Use visual methods such as box plots or scatter plots to better understand the nature of the outliers and determine if they are genuine or due to errors.

### Missing Values in FAO - Hybrid FAO:
- **Replace the NaNs with the average FAO value for the Geozone**: Consider implementing this approach after verifying that it's appropriate for the context and distribution of your data. Ensure that it won't introduce any bias or misrepresentation.

### Feature 'K_kg/ga' Investigation:
- **Further investigate the 'K_kg/ga' feature to determine if rescaling to a 0-1 range is necessary**: Analyze the current scale and distribution of the 'K_kg/ga' values to decide if rescaling would make the data more consistent and interpretable, especially if it's significantly different from other features.

### WEEK_RAINFALL Outlier Investigation:
- **Further investigate the 'WEEK_RAINFALL' feature for outliers**: Given the noted discrepancies between the mean and median and the large distance between the 99.5 and 100 percentiles, a deeper investigation into potential outliers and their impact on the analysis is recommended.

### Imputation Strategy for Missing Values:
- **Determine imputation strategies for missing values**: Aside from the suggested imputation for the FAO feature, consider appropriate strategies for other features with missing values. This might include deletion, imputation with mean/median/mode, or more complex methods like predictive modeling based on other features.
